### ISSUE TITLE
OpenMM ffxml water model templates cannot have H-H bonds

### ISSUE THREAD

**Issue by jchodera** [choderalab/openmm-forcefields#56]
We will need to regenerate the CHARMM water models to ensure they do not contain H-H bonds.

cc: https://github.com/pandegroup/openmm/issues/1939#issuecomment-359130400
cc: https://github.com/ParmEd/ParmEd/issues/941

**Comment by jchodera** [comment]
@peastman : I was trying to figure out how this issue slipped through since we actually *test* assigning parameters to a PDB file containing waters in [this test](https://github.com/ParmEd/ParmEd/blob/master/test/test_parmed_openmm.py#L612-L633), and I realized that this test must be passing because the default CHARMM [water parameter file](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/toppar/toppar_water_ions.str) actually contains both [TIP3P](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/toppar/toppar_water_ions.str#L65-L75) (which contained the artificial H-H bond) and [the MMFF water model](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/toppar/toppar_water_ions.str#L77-L87) (which does not have an H-H bond).

Once we remove the H-H bond from TIP3P, the problem will be that this file will now contain two identical templates. That means we will need some degree of manual processing, which I'd prefer to program via the YAML file. 

What should we do here? My thought is to specify when certain residues are to be excluded from reading, so we can program this to produce two separate files (`waters_ions_default.xml` and `waters_ions_mmff.xml`):

```yaml
- sourcePackage: charmm36
  sourcePackageVersion: toppar_c36_aug15.tgz
- Destination: waters_ions_default.xml
  Source:
    stream:
     - toppar/toppar_water_ions.str
    exclude_residues:
     - TP3M
  References:
   tip3p:
     - >-
       W.L. Jorgensen; J.Chandrasekhar; J.D. Madura; R.W. Impey; M.L. Klein; "Comparison of simple potential functions for
       simulating liquid water", J. Chem. Phys. 79 926-935 (1983).
   ions:
     - >-
       Beglov, D. and Roux, B., Finite Representation of an Infinite Bulk System: Solvent Boundary Potential for
       Computer Simulations, Journal of Chemical Physics, 1994, 100: 9050-9063
     - zinc:
       - >-
         Stote, R.H. and Karplus, M. Zinc Binding in Proteins and Solution: A Simple but Accurate Nonbonded
         Representation, PROTEINS: Structure, Function, and Genetics 23:12-31 (1995)
     - cations:
       - >-
         Won, Y., "Force Field for Monovalent, Divalent, and Trivalent Cations Developed under the Solvent Boundary
         Potential," Journal of Physical Chemistry A., 116: 11763-11767, 2012.
- Destination: waters_ions_mmff.xml
  Source:
    stream:
     - toppar/toppar_water_ions.str
    exclude_residues:
     - TIP3
  References:
   MMFF:
     - >-
       [ref goes here]
   ions:
     - >-
       Beglov, D. and Roux, B., Finite Representation of an Infinite Bulk System: Solvent Boundary Potential for
       Computer Simulations, Journal of Chemical Physics, 1994, 100: 9050-9063
     - zinc:
       - >-
         Stote, R.H. and Karplus, M. Zinc Binding in Proteins and Solution: A Simple but Accurate Nonbonded
         Representation, PROTEINS: Structure, Function, and Genetics 23:12-31 (1995)
     - cations:
       - >-
         Won, Y., "Force Field for Monovalent, Divalent, and Trivalent Cations Developed under the Solvent Boundary
         Potential," Journal of Physical Chemistry A., 116: 11763-11767, 2012.
```

**Comment by peastman** [comment]
That looks reasonable.

**Comment by jchodera** [comment]
@peastman : What would you think about having OpenMM's [`ForceField`](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py) class check for duplicate residue templates on reading of an ffxml file? It would make it much easier for us to catch problems earlier this way.

**Comment by peastman** [comment]
Duplicate in what sense?  In this case, there were two templates with different names and different bonds.  How would it have recognized there was a problem with them?

It already checks for multiple templates with the same name, and throws an exception if they have the same override level.

**Comment by jchodera** [comment]
> Duplicate in what sense? In this case, there were two templates with different names and different bonds. How would it have recognized there was a problem with them?

When we remove the `H-H` bond, we end up with `TIP3` and `TP3M` templates being identical. I was hoping for a simple way of checking that the resulting ffxml files were well-formed.

For now, for testing/validation, I'm loading in a water box and attempting to do a `createSystem()` from that topology. However, testing this way is more difficult since I need a PDB file tailored to each xml file I'm generating---checking for duplicates on loading the ffxml would be much more convenient and general.

**Comment by peastman** [comment]
If multiple templates match a residue, `createSystem()` throws an exception.

**Comment by jchodera** [comment]
Yes, but in order to use `createSystem()` to test whether you have multiple templates, you need to load a PDB file containing residues that match the templates you want to check for duplicates. This requires a PDB file tailored to each ffxml file you want to check for duplicate residues.

Instead, I'm suggesting we have `ForceField` check for duplicates when the ffxml file is loaded. That will make it very easy to tell if there are duplicate residues by simply loading the ffxml file into `ForceField()`.

**Comment by jchodera** [comment]
Also, to clarify what the original problem was here: The original `waters_ions_default.xml` contained two water topologies: One with the H-H bond, and one without, with identical parameters. That's why this slipped through testing: The one without the H-H bond allowed us to parameterize protein+solvent systems just fine.

Now that we are removing H-H bonds for waters, we have duplicate residues, and need to be more careful about how we check for duplicate residues so we don't run into more problems in the future.

**Comment by jchodera** [comment]
This was addressed by #58
